File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step2/F/F.cpp
// author: Danila "akshin_" Axyonov

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int INF = (int)2e9 + 29;



class DSU {
private:
    vector<int> p;
    vector<int> r;
    int size;

public:
    void init(int orsize) {
        this->size = orsize;
        this->p.resize(this->size);
        this->r.resize(this->size);
        this->clear();
        return;
    }

    void clear() {
        for (int i = 0; i < size; ++i) {
            this->p[i] = i;
            this->r[i] = 0;
        }
        return;
    }

    int get(int a) {
        if (this->p[a] != a)
            this->p[a] = this->get(this->p[a]);
        return this->p[a];
    }

    void unite(int a, int b) {
        a = this->get(a);
        b = this->get(b);
        if (a == b)
            return;
        if (this->r[a] == this->r[b])
            ++this->r[a];
        if (this->r[a] > this->r[b])
            this->p[b] = a;
        else // this->r[a] < this->r[b]
            this->p[a] = b;
        return;
    }
};



struct Rib {
    int u;
    int v;
    int w;
};



void solve();

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    solve();
    
    return 0;
}

void solve() {
    int n,
        m;
    cin >> n >> m;
    vector<Rib> ribs(m);
    for (int i = 0; i < m; ++i) {
        cin >> ribs[i].u; --ribs[i].u;
        cin >> ribs[i].v; --ribs[i].v;
        cin >> ribs[i].w;
    }
    sort(ribs.begin(), ribs.end(), [](Rib r1, Rib r2) { return r1.w < r2.w; });

    DSU dsu;
    dsu.init(n);
    int min_d = INF;
    for (int i = 0; i < m; ++i) {
        dsu.clear();
        int _min = INF,
            _max = -INF,
            cnt = 0;
        for (int j = i; j < m and cnt < n - 1; ++j) {
            if (dsu.get(ribs[j].u) == dsu.get(ribs[j].v))
                continue;
            dsu.unite(ribs[j].u, ribs[j].v);
            _min = min(_min, ribs[j].w);
            _max = max(_max, ribs[j].w);
            ++cnt;
        }
        if (cnt == n - 1)
            min_d = min(min_d, _max - _min);
    }
    
    if (min_d == INF) // there is no MST (minimal spanning tree)
        cout << "NO\n";
    else // there is at least one MST
        cout << "YES\n" << min_d << '\n';

    return;
}

File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step2/F/F.py
import operator as op
INF = 2 * 10 ** 9 + 29

class DSU:
    def __init__(self, orsize: int) -> None:
        self.size = orsize
        self.clear()
        return
    
    def clear(self) -> None:
        self.p = list(range(self.size))
        self.r = [0] * self.size
        return
    
    def get(self, a: int) -> int:
        if self.p[a] != a:
            self.p[a] = self.get(self.p[a])
        return self.p[a]
    
    def unite(self, a: int, b: int) -> None:
        a = self.get(a)
        b = self.get(b)
        if a == b:
            return
        if self.r[a] == self.r[b]:
            self.r[a] += 1
        if self.r[a] > self.r[b]:
            self.p[b] = a
        else: # self.r[a] < self.r[b]
            self.p[a] = b
        return

def main() -> None:
    n, m = map(int, input().split())
    ribs = []
    for _ in range(m):
        u, v, w = map(int, input().split())
        ribs.append((u - 1, v - 1, w))
    ribs.sort(key=op.itemgetter(-1))
    
    dsu = DSU(n)
    min_d = INF
    for i in range(m):
        dsu.clear()
        _min = INF
        _max = -INF
        cnt = 0
        for j in range(i, m):
            if cnt == n - 1:
                break
            if dsu.get(ribs[j][0]) == dsu.get(ribs[j][1]):
                continue
            dsu.unite(ribs[j][0], ribs[j][1])
            _min = min(_min, ribs[j][-1])
            _max = max(_max, ribs[j][-1])
            cnt += 1
        if cnt == n - 1:
            min_d = min(min_d, _max - _min)
    
    if min_d == INF: # there is no MST (minimal spanning tree)
        print('NO')
    else: # there is at least one MST
        print(f'YES\n{min_d}')

    return

if __name__ == '__main__':
    main()
